HLTrigger/Muon/plugins/HLTMuonTrkL1TkMuFilter.cc

```diff
@@ -0,0 +1,137 @@
+  m_previousCandTag = iConfig.getParameter<edm::InputTag>("previousCandTag");
+  m_previousCandToken = consumes<trigger::TriggerFilterObjectWithRefs>(m_previousCandTag);
+  m_minTrkHits = iConfig.getParameter<int>("minTrkHits");
+  m_minMuonHits = iConfig.getParameter<int>("minMuonHits");
+  m_minMuonStations = iConfig.getParameter<int>("minMuonStations");
+  m_maxNormalizedChi2 = iConfig.getParameter<double>("maxNormalizedChi2");
+  m_minPt = iConfig.getParameter<double>("minPt");
+  m_minN = iConfig.getParameter<unsigned int>("minN");
+  m_maxAbsEta = iConfig.getParameter<double>("maxAbsEta");
+}
+
+void HLTMuonTrkL1TkMuFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  makeHLTFilterDescription(desc);
+  desc.add<edm::InputTag>("inputMuonCollection", edm::InputTag(""));
+  desc.add<edm::InputTag>("inputCandCollection", edm::InputTag(""));
+  desc.add<edm::InputTag>("previousCandTag", edm::InputTag(""));
+  desc.add<int>("minTrkHits", -1);
+  desc.add<int>("minMuonHits", -1);
+  desc.add<int>("minMuonStations", -1);
+  desc.add<double>("maxNormalizedChi2", 1e99);
+  desc.add<unsigned int>("trkMuonId", 0);
+  desc.add<double>("minPt", 24);
+  desc.add<unsigned int>("minN", 1);
+  desc.add<double>("maxAbsEta", 1e99);
+  descriptions.add("hltMuonTrkL1TkMuFilter", desc);
+}
+
+bool HLTMuonTrkL1TkMuFilter::hltFilter(edm::Event& iEvent,
+                                       const edm::EventSetup& iSetup,
+                                       trigger::TriggerFilterObjectWithRefs& filterproduct) const {
+  edm::Handle<reco::MuonCollection> muons;
+  iEvent.getByToken(m_muonsToken, muons);
+  edm::Handle<reco::RecoChargedCandidateCollection> cands;
+  iEvent.getByToken(m_candsToken, cands);
+  if (saveTags())
+    filterproduct.addCollectionTag(m_candsTag);
+  if (cands->size() != muons->size())
+    throw edm::Exception(edm::errors::Configuration)
+        << "Both input collection must be aligned and represent same physical muon objects";
+
+  edm::Handle<trigger::TriggerFilterObjectWithRefs> previousLevelCands;
+  std::vector<l1t::TkMuonRef> vl1cands;
+  std::vector<l1t::TkMuonRef>::iterator vl1cands_begin;
+  std::vector<l1t::TkMuonRef>::iterator vl1cands_end;
+
+  bool check_l1match = true;
+  if (m_previousCandTag == edm::InputTag(""))
+    check_l1match = false;
+  if (check_l1match) {
+    iEvent.getByToken(m_previousCandToken, previousLevelCands);
+    previousLevelCands->getObjects(trigger::TriggerL1TkMu, vl1cands);
+    vl1cands_begin = vl1cands.begin();
+    vl1cands_end = vl1cands.end();
+  }
+
+  std::vector<unsigned int> filteredMuons;
+  for (unsigned int i = 0; i < muons->size(); ++i) {
+    const reco::Muon& muon(muons->at(i));
+    // check for dR match to L1 muons
+    if (check_l1match) {
+      bool matchl1 = false;
+      for (auto l1cand = vl1cands_begin; l1cand != vl1cands_end; ++l1cand) {
+        if (deltaR(muon, **l1cand) < 0.3) {
```

Comment:
this could be simplified as
```suggestion
      for (auto const& l1cand: vl1cands) {
        if (deltaR(muon, *l1cand) < 0.3) {
```

Reply:
`deltaR2`